### ISSUE TITLE
1.16.2: API: Loading schematics gives "EOFException: Stream ended prematurely"

### ISSUE THREAD

**Issue by kangarko** [IntellectualSites/FastAsyncWorldEdit#605]
# Bug report for FastAsyncWorldEdit 1.16.2

**FastAsyncWorldEdit Configuration Files:** 
Default settings: https://athion.net/ISPaster/paste/view/ea487a4624434c82a1bdfdd720b6ebc6

**FastAsyncWorldEdit Version Number:** 
https://i.imgur.com/zcHsEJ1.png

**Spigot/Paper Version Number:** 
https://i.imgur.com/086drqP.png

**Description of the problem:**
Calling ClipboardReader#read() produces EOFException: Stream ended prematurely error when using FastAsyncWorldEdit. **Using the vanilla WorldEdit, no error is produced**.

**How to replicate**:
Here is our full code handling loading/saving/pasting schematics: https://pastebin.com/ARJPPZMD
Line 95 produces this error: https://pastebin.com/zuUNqmp9

**Plugins being used on the server:**
https://i.imgur.com/ZAGS6cA.png

**Checklist**:

- [x] I included all information required in the sections above
- [x] I made sure there are no duplicates of this report [(Use Search)](https://github.com/IntellectualSites/FastAsyncWorldEdit/issues?q=is%3Aissue)
- [x] I made sure I am using an up-to-date version of [FastAsyncWorldEdit for 1.16.2](https://ci.athion.net/job/FastAsyncWorldEdit-1.16/)
- [x] I made sure the bug/error is not caused by any other plugin

**Comment by aurorasmiles** [comment]
It could be that cf750ba00c81015f2563c1d4a8f6a1e991dfc7c6 fixed this, could you test again?

**Comment by kangarko** [comment]
@LeehamElectronics could you please help testing this one? Just use their latest Jenkins build

**Comment by LeehamElectronics** [comment]
> 
> 
> @LeehamElectronics could you please help testing this one? Just use their latest Jenkins build

Yep ill do that now! Thanks

**Comment by LeehamElectronics** [comment]
So this isn't a development build? Do I have to compile it into a jar file?

**Comment by aurorasmiles** [comment]
You can just use the build on our jenkins.

**Comment by LeehamElectronics** [comment]
Well the plugin no longer returns any errors when taking snapshots (aka saving region as a SCHEM) however when I restore the arena (region) to its initial or damaged state, simply nothing happens, the plugin does not change any of the blocks at all, and there are no errors in the console, any ideas @kangarko  ?
I will try and put everything in debug mode to see if there are any other error messages.

**Comment by LeehamElectronics** [comment]
Nope there was no error messages even with debug on, however it looks like CoreArena may not be detecting Worldedit (I'm not really sure) because I see this message as CoreArena starts up:
`Warning: WorldEdit is required for procedural damage and arena snapshots.`
Is that normal? Anyway I'm out of ideas, let me know if there's anything else I can try.
Cheers

**Comment by kangarko** [comment]
@LeehamElectronics apology, that was a problem on my end. 

I fixed that but testing FAWE build 378 I am still getting that error:

```
[19:47:18 INFO]: CoreArena 2.7.13 encountered an EOFException! Please check your error.log and report this issue with the information in that file.
[19:47:18 WARN]: org.mineacademy.fo.exception.FoException: Report: Failed to load INITIAL template for Test, see below and report (make sure you have the latest WorldEdit)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.SimpleSnapshotWorldEdit.load(SimpleSnapshotWorldEdit.java:108)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.SimpleSnapshotWorldEdit.loadAll(SimpleSnapshotWorldEdit.java:62)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.SimpleSnapshotWorldEdit.<init>(SimpleSnapshotWorldEdit.java:57)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.arena.SimpleArena.<init>(SimpleArena.java:112)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.arena.FeatureArena.<init>(FeatureArena.java:64)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.arena.MobArena.<init>(MobArena.java:21)
[19:47:18 WARN]: 	at org.mineacademy.game.manager.SimpleArenaManager.loadFromFile(SimpleArenaManager.java:73)
[19:47:18 WARN]: 	at org.mineacademy.game.manager.SimpleArenaManager.loadSavedArenas(SimpleArenaManager.java:46)
[19:47:18 WARN]: 	at org.mineacademy.corearena.CoreArenaPlugin.startReloadable(CoreArenaPlugin.java:192)
[19:47:18 WARN]: 	at org.mineacademy.corearena.CoreArenaPlugin.onPluginStart(CoreArenaPlugin.java:174)
[19:47:18 WARN]: 	at org.mineacademy.fo.plugin.SimplePlugin.onEnable(SimplePlugin.java:294)
[19:47:18 WARN]: 	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:263)
[19:47:18 WARN]: 	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:380)
[19:47:18 WARN]: 	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:483)
[19:47:18 WARN]: 	at org.bukkit.craftbukkit.v1_16_R2.CraftServer.enablePlugin(CraftServer.java:501)
[19:47:18 WARN]: 	at org.bukkit.craftbukkit.v1_16_R2.CraftServer.enablePlugins(CraftServer.java:415)
[19:47:18 WARN]: 	at net.minecraft.server.v1_16_R2.MinecraftServer.loadWorld(MinecraftServer.java:468)
[19:47:18 WARN]: 	at net.minecraft.server.v1_16_R2.DedicatedServer.init(DedicatedServer.java:237)
[19:47:18 WARN]: 	at net.minecraft.server.v1_16_R2.MinecraftServer.w(MinecraftServer.java:939)
[19:47:18 WARN]: 	at net.minecraft.server.v1_16_R2.MinecraftServer.lambda$a$0(MinecraftServer.java:177)
[19:47:18 WARN]: 	at java.lang.Thread.run(Thread.java:748)
[19:47:18 WARN]: Caused by: java.io.EOFException: Stream ended prematurely
[19:47:18 WARN]: 	at net.jpountz.lz4.LZ4BlockInputStream.readFully(LZ4BlockInputStream.java:212)
[19:47:18 WARN]: 	at net.jpountz.lz4.LZ4BlockInputStream.refill(LZ4BlockInputStream.java:142)
[19:47:18 WARN]: 	at net.jpountz.lz4.LZ4BlockInputStream.read(LZ4BlockInputStream.java:94)
[19:47:18 WARN]: 	at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.readCombined(SchematicReader.java:175)
[19:47:18 WARN]: 	at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.readwrite(SchematicReader.java:196)
[19:47:18 WARN]: 	at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:268)
[19:47:18 WARN]: 	at com.sk89q.worldedit.extent.clipboard.io.ClipboardReader.read(ClipboardReader.java:59)
[19:47:18 WARN]: 	at com.sk89q.worldedit.extent.clipboard.io.ClipboardReader.read(ClipboardReader.java:46)
[19:47:18 WARN]: 	at org.mineacademy.game.impl.SimpleSnapshotWorldEdit.load(SimpleSnapshotWorldEdit.java:95)
[19:47:18 WARN]: 	... 20 more
```

This is the code that causes it: https://i.imgur.com/088XXOB.png

@aurorasmiles what can I do here to fix this please? Thanks!

**Comment by LeehamElectronics** [comment]
@kangarko  hmm I have tried twice on the latest versions of FAWE and CoreArena and I am not getting any errors, instead nothing happens at all, no blocks are created or removed. Not sure why though.

**Comment by dordsor21** [comment]
Try using the latest build

**Comment by LeehamElectronics** [comment]
> 
> 
> Try using the latest build

Just tried it today and I am still getting the same issue, perhaps @kangarko could give it a try too.
Thanks!

**Comment by kangarko** [comment]
@LeehamElectronics just post the error here, I have not changed any code and I am only using WorldEdit native API, that should in theory work with FAWE

**Comment by LeehamElectronics** [comment]
Sorry @kangarko I am not getting any errors at all, I have procedural damage enabled and debug mode is on, yet there are no error messages when loading snapshots, instead, nothing happens, CoreArena says that the snapshot loaded and saved successfully but no blocks are removed / added.
Am I doing something wrong?

**Comment by LeehamElectronics** [comment]
This also happens when using normal world edit as well, snapshots don't do anything. 
https://ibb.co/3MMH9cm

**Comment by LeehamElectronics** [comment]
**Sorry let me clarify,** when using _normal World Edit_ everything works fine with CoreArena, when **using FAWE I do NOT receive any error messages and NO blocks are placed or removed** when loading snapshots. In my previous comment you can see the image that says that the snapshots have not been set, even though I have clearly pressed the save snapshot button. I also receive this message on startup for some reason:
`[12:49:28 INFO]: Warning: WorldEdit is required for procedural damage and arena snapshots.`

In summary, it seems that when using FAWE, CoreArena is not saving schematics or loading them at all, and it is not generating any error messages either. Sorry for being so upfront in my comment, I am just trying to word it as clearly as possible.
Thanks in advanced!

**Comment by LeehamElectronics** [comment]
I deleted my previous messages as I made a mistake, using the latest FAWE and latest CoreArena I get the following error when saving snapshots:
```
[13:11:10 ERROR]: EOFException read in schematic. Did you give the schematic the wrong extension?
[13:11:10 ERROR]: We will attempt to rectify your mistake for you and load the schematic assuming it is named .schem not .schematic
[13:11:10 WARN]: java.io.EOFException: Stream ended prematurely
[13:11:10 WARN]:        at net.jpountz.lz4.LZ4BlockInputStream.readFully(LZ4BlockInputStream.java:212)
[13:11:10 WARN]:        at net.jpountz.lz4.LZ4BlockInputStream.refill(LZ4BlockInputStream.java:142)
[13:11:10 WARN]:        at net.jpountz.lz4.LZ4BlockInputStream.read(LZ4BlockInputStream.java:94)
[13:11:10 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.readCombined(SchematicReader.java:184)
[13:11:10 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.readwrite(SchematicReader.java:205)
[13:11:10 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.readInternal(SchematicReader.java:296)
[13:11:10 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:231)
[13:11:10 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.ClipboardReader.read(ClipboardReader.java:59)
[13:11:10 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.ClipboardReader.read(ClipboardReader.java:46)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.ei.a(SourceFile:95)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.ei.b(SourceFile:62)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.ei.a(SourceFile:120)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.fx.e(SourceFile:33)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.fb$a.a(SourceFile:132)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.aF.a(SourceFile:780)
[13:11:10 WARN]:        at org.mineacademy.corearena.i.aG.onMenuClick(SourceFile:75)
[13:11:10 WARN]:        at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor1002.execute(Unknown Source)
[13:11:10 WARN]:        at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69)
[13:11:10 WARN]:        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80)
[13:11:10 WARN]:        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70)
[13:11:10 WARN]:        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:607)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.PlayerConnection.a(PlayerConnection.java:2427)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.PacketPlayInWindowClick.a(SourceFile:32)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.PacketPlayInWindowClick.a(SourceFile:10)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.PlayerConnectionUtils.lambda$ensureMainThread$1(PlayerConnectionUtils.java:23)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.TickTask.run(SourceFile:18)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.IAsyncTaskHandler.executeTask(IAsyncTaskHandler.java:136)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.IAsyncTaskHandler.executeNext(IAsyncTaskHandler.java:109)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.MinecraftServer.ba(MinecraftServer.java:1135)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.MinecraftServer.executeNext(MinecraftServer.java:1128)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:119)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.MinecraftServer.sleepForTick(MinecraftServer.java:1089)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.MinecraftServer.w(MinecraftServer.java:1003)
[13:11:10 WARN]:        at net.minecraft.server.v1_16_R2.MinecraftServer.lambda$a$0(MinecraftServer.java:177)
[13:11:10 WARN]:        at java.lang.Thread.run(Unknown Source)
[13:11:10 INFO]: CoreArena 2.7.14 encountered a NullPointerException! Please check your error.log and report this issue with the information in that file.
[13:11:10 INFO]: java.lang.NullPointerException
[13:11:10 INFO]: at org.mineacademy.corearena.i.ei.a(SourceFile:101)
[13:11:10 INFO]: at org.mineacademy.corearena.i.ei.b(SourceFile:62)
[13:11:10 INFO]: at org.mineacademy.corearena.i.ei.a(SourceFile:120)
[13:11:10 INFO]: at org.mineacademy.corearena.i.fx.e(SourceFile:33)
[13:11:10 INFO]: at org.mineacademy.corearena.i.fb$a.a(SourceFile:132)
[13:11:10 INFO]: at org.mineacademy.corearena.i.aF.a(SourceFile:780)
[13:11:10 INFO]: at org.mineacademy.corearena.i.aG.onMenuClick(SourceFile:75)
[13:11:10 INFO]: !-----------------------------------------------------!
[13:11:10 INFO]: Error clicking in menu fb{}
[13:11:10 INFO]: !-----------------------------------------------------!
```

**Comment by kangarko** [comment]
Yeah same source code, you can see the source with line 95 in the top post.

**Comment by LeehamElectronics** [comment]
I'll give $15 AUD via PayPal to anyone who can get this bug fixed with FAWE, it's really important for me as CoreArena is the most popular mini-game on my server.
Thanks!

**Comment by aurorasmiles** [comment]
Just to check a theory I have, could you provide the schematic in question? @LeehamElectronics

**Comment by aurorasmiles** [comment]
@kangarko It seems like your pastebins have expired, could you please upload them somewhere where they wont get removed?

**Comment by LeehamElectronics** [comment]
Thanks @aurorasmiles here is a link to the schematic file generated by CoreArena using the latest version and latest dev build of FAWE:
https://www.dropbox.com/sh/sijpnj8t3rriyeq/AACwVG9LY7wDtKkfjLzrDxTBa?dl=0
Thanks in advanced!

**Comment by aurorasmiles** [comment]
I believe this issue is rather simple.
@kangarko You save the schematics in Sponge Format, but as a .schematic file, instead of .schem
WorldEdit is able to convert that on the fly, while FAWE isn't. Please save the schematic in the correct format (name.schem) when using the Sponge Format, that should solve this issue

**Comment by kangarko** [comment]
Pardon me @aurorasmiles, but there are only two options: https://i.imgur.com/krWERjj.png

Setting it to the other option, MCEDIT, produces: https://i.imgur.com/W4iaOoT.png

Setting it back to SPONGE produces this on load: https://i.imgur.com/eVQnyvn.png

**Comment by weaondara** [comment]
line 284: cache.put(stage, new File(CoreArenaPlugin.getInstance().getDataFolder().getParent(), getSchematicFolder() + stage.getFileName().replace("{arena}", getArena().getName()) + ".schematic"));
there

**Comment by aurorasmiles** [comment]
Change .schematic to .schem in that line @kangarko

**Comment by SirYwell** [comment]
You can also use the SPONGE_SCHEMATIC for the reader format directly, then the file extension doesn't matter

**Comment by SirYwell** [comment]
Yes, FAWE uses a more optimized Clipboard implementation here. What's the reason you need a BlockArrayClipboard for?

**Comment by kangarko** [comment]
We no longer need it, I have updated the code and IT WORKS!

Just a glitch though, the terrain is not rendered with lightning: https://i.imgur.com/GgXsHPP.png -- is there a method we can call to force update it?

Current code: https://pastebin.com/cqcUQBR9

**Comment by aurorasmiles** [comment]
You can use FaweAPI.fixLighting for that

**Comment by kangarko** [comment]
Thank you, finally got it all working! <3
